[ArticulatedSystemPlugin] New example

EulalieCoevoet opened the pull request:
Adds an example to the plugin ArticulatedSystemPlugin. 

![image](https://user-images.githubusercontent.com/12150715/159580224-bb724ee4-f469-4697-b373-e6e22c0b4a86.png)


______________________________________________________

By submitting this pull request, I acknowledge that  
**I have read, understand, and agree [SOFA Developer Certificate of Origin (DCO)](https://github.com/sofa-framework/sofa/blob/master/CONTRIBUTING.md#sofa-developer-certificate-of-origin-dco)**.
______________________________________________________

**Reviewers will merge this pull-request only if**  
- it builds with SUCCESS for all platforms on the CI.
- it does not generate new warnings.
- it does not generate new unit test failures.
- it does not generate new scene test failures.
- it does not break API compatibility.
- it is more than 1 week old (or has fast-merge label).

fredroy committed:
move files and change cmake files accordingly

fredroy committed:
clean _API, etc

fredroy committed:
add missing licence header in init/config.h

fredroy committed:
update cmake for depreciation, redirection and add compat

fredroy committed:
update component change

fredroy committed:
fix multithreading

fredroy committed:
add dates for deprecations warnings

fredroy committed:
move SimpleTesselated and Mesh2Point topomapping

fredroy committed:
clean, compat for simpletesselated and mesh2point

fredroy committed:
fix sofacuda (include for what you use)

fredroy committed:
create solidmechanics module (and its compat layer)

fredroy committed:
create tensormass module (and its compat layer)

fredroy committed:
fix sofacuda, as the compat files are not usable for specialization

fredroy committed:
add forgotten redirection for deprecated sofanonuniform module

fredroy committed:
add ray-tri intersection test in geometry::Triangle + array version for normal

fredroy committed:
add unittest for intersection (+ expand tests for normal to arrays)

fredroy committed:
replace use of RayTriangleIntersection with the static one from geometry

fredroy committed:
deprecate RayTriangleIntersection class (and remove compilation unit)

fredroy committed:
fix issue with isnan (just use std::isnan() directly)

fredroy committed:
create module container

fredroy committed:
update component change

fredroy committed:
Apply suggestions from code review (typo, static constexpr, var replacements)

Co-authored-by: Alex Bilger <alxbilger@users.noreply.github.com>

fredroy committed:
apply suggestion on template, and noinit vec

fredroy committed:
fix test of isnan on the inv_det

fredroy committed:
split mappings between linear and nonlinear ones

fredroy committed:
clean nonlinear mappings, does not need sofa.comp.topology anymore

fredroy committed:
update compat files, reflecting split

fredroy committed:
update component change list

fredroy committed:
fix typos and tests

fredroy committed:
fix sofacuda

fredroy committed:
create sub module mappedmatrix

fredroy committed:
update component change list

fredroy committed:
create new module mechanicalload

fredroy committed:
add compat layer

fredroy committed:
fix sofacuda

fredroy committed:
fix compilation for gcc/clang

fredroy committed:
restore PenalityContactForceField, which should be in contacts

maddin200 committed:
[modules] remove shadow variable

fredroy committed:
move triangleoctree utility in helper + compat

fredroy committed:
apply move and namespace change

fredroy committed:
remove incorrect (and weird??) inclusion of defaulttype/fwd.h in drawtool

EulalieCoevoet committed:
[ArticulatedSystemPlugin] new example

alxbilger committed:
[SofaCore] Virtual inheritance of BaseObject (#2799)

fredroy committed:
replace macro bb_min/max with stl function std::min/max

fredroy committed:
replace cube size macro with a static constant

fredroy committed:
add deleted function to keep the removed function

EulalieCoevoet committed:
[SofaCore] restore default name for python (#2801)

fredroy committed:
move hyperelastic materials into a nested folder

fredroy committed:
rename state to statecontainer

fredroy committed:
update sofacuda

fredroy committed:
fix compilation

alxbilger committed:
[Sofa.LinearAlgebra] Deprecate unused EigenMatrixManipulator (#2793)

* [Sofa.LinearAlgebra] Deprecate unused EigenMatrixManipulator

* Fix typo

* fix for gcc

* fix clang

Co-authored-by: Frederick Roy <frederick.roy@inria.fr>

fredroy committed:
[sofaProjectExample] Resuscitate application with gui and simulation (cpp) (#2792)

EulalieCoevoet committed:
[SofaGuiCommon] restores the use of argv with runsofa (#2802)

alxbilger committed:
[Scenes] Fix plugins in a couple of scenes (#2810)

* SofaNewmat is no longer in the core

* Fix typo

fredroy committed:
remove dubious optimization

alxbilger committed:
[Scenes] Remove missing files from the scene list (#2809)

alxbilger committed:
[Regression] Test linear solvers for regression (#2776)

* [Regression] Test linear solvers for regression

* Update because of renamed

* Rename to reduce paths length

* Another try to reduce paths length

fredroy committed:
update component change

alxbilger committed:
[LinearSolver.Iterative] Clean floating point types in CG (#2808)

fredroy commented:
I see that this PR is adding ~10MB of data (from the stl files).
I think we should wait for this plugin to be externalized (on an other repo) to merge this example, otherwise the history of this repo will contain those files, and the SOFA repo history is already so cluttered 🤪
what do you think @hugtalbot  ?

fredroy committed:
fix sofacuda

EulalieCoevoet committed:
[ArticulatedSystem] removes unused stl files

EulalieCoevoet committed:
[ArticulatedSystem] cleaning

EulalieCoevoet commented:
I've just cleaned the repo, removed unused mesh files.
It's 1.7MB of stl files now. Is it still too much ? I could also simplify the meshes but, yeah... it won't be as "pretty".

fredroy committed:
fix sofacuda (again)

fredroy committed:
fix test compilation

epernod committed:
[SofaGuiQt] Fix english spelling: AdvancedTimer instead of AdvanceTimer

fredroy committed:
update component change list

alxbilger committed:
[LMConstraint] Remove LMConstraint folder as it is an external plugin (#2828)

See https://github.com/sofa-framework/LMConstraint and https://github.com/sofa-framework/sofa/pull/2508

fredroy committed:
fix scene (closing Node) (#2831)

epernod committed:
[Scenes] Fix ProjectToPointconstraint scene by using MeshSpringForceField (#2827)

* [scenes] Fix ProjectToPointconstraint scene by using springs instead of FEM method

* fix indent

* Update ProjectToPointConstraint.scn

* Update ProjectToPointConstraint.scn

alxbilger committed:
[SofaEulerianFluid] Move examples into plugin (#2833)

fredroy committed:
Remove annoying unused warning in StateAccessor (#2835)

alxbilger committed:
[Component.IO.Mesh] Fix substring comparison (#2837)

Fixes #2576

alxbilger committed:
[SofaBoundaryCondition] Remove duplicated code (#2830)

Code duplicated from the base class constructor, therefore called twice

epernod committed:
[SofaSimpleFEM] Create a TriangleFEMUtils class to factorise Triangle/Triangular[FEMForceField] code (#2287)

* [SofaSimpleFem] Add class TriangleFEMUtils to factorise some code

* ﻿[SofaMiscFem] Start using TriangleFEMUtils in Triangular and TriangleFEMForcefield. Factorise ComputeRotationLarge

* [SofaMiscFem] Factorise Triangular and TriangleFEMForcefield using TriangleFEMUtils. Factorise method ComputeStrainDisplacement

* [SofaMiscFem] modify ComputeStrainDisplacementLocal to take only pB and pC coordinates

* [SofaMiscFem] Improve TriangularFEM initSmall

* [SofaMiscFem] Factorise Triangular and TriangleFEMForcefield using TriangleFEMUtils. Factorise computeStrain and computeStress methods

* [SofaMiscFem] Factorise Triangular and TriangleFEMForcefield using TriangleFEMUtils. Factorise method computeDisplacementSmall and computeDisplacementLarge

* [sofaMiscFem] clean TriangularFEM computeForce, accumulateForceSmall and accumulateForceLarge to match TriangleFEMForceField

* ﻿[SofaMiscFem] Clean TriangularFEMForceField applystiffnessSmall and Large

* Put method in constExpr

* [SofaMiscFem] Merge accumulateForce and ComputeForce to avoid double triangleInfo access

* Optimise TriangularFEMForceField

* ﻿[sofaMiscFem] Update Triangle and TriangularFEM components to loop on element directly inside accumulateForce and applyStiffness  Small/Large

* [SofaMiscFem] Fix TriangleFEMForceField init

* [SofaMiscFem] Move all comments out from code. Pass all method as const

* [SofaMiscFem] Add determinant check not null

* [SofaMiscFem] Do not compute Jt in optimised version of computeStrain

* [SofaMiscFem] remove loop on getTriangle and multiple access of Veccord in a same loop. Those methods seems not to be optimised by compiler.

* Apply suggestions from code review

Co-authored-by: Alex Bilger <alxbilger@users.noreply.github.com>

* [src] Set NoInit

* [SofaMiscFem] Fix methods to use Real instead of SReal

* [SofaMiscFem] use TriangleUtils as object instead of pointer

* [SofaMiscFem] Add try/catch exception on computeStrainDisplacement determinant computation.

* Update TriangleFEMUtils.inl

* [FEM] Fix compilation by moving TriangleFEMUtils in the new directory

* test for debug

* test for debug

* test for debug

* test for debug

* test for debug

* [FEM_elastic] use fabs instead of abs, even if it should be the same.

* change test to investigate

Co-authored-by: Alex Bilger <alxbilger@users.noreply.github.com>

alxbilger committed:
[examples] Remove scene which should be in LMConstraint (#2834)

fredroy committed:
[Sofa.Component.Diffusion][Sofa.Component.Mass] Fix diffusion (#2798)

* add template for geometry info

* apply to sofacuda

* fix templates

* clean and remove option on 1d diffusion

* apply to scene and tests

* Update TetrahedronDiffusionFEMForceField.scn

* remove useless mult by 1.0

Co-authored-by: Alex Bilger <alxbilger@users.noreply.github.com>

* replace direct ptr with Links

* remove specific 1d branch

* Update Component/Mass/src/sofa/component/mass/DiagonalMass.inl (change msg_info to warn)

Co-authored-by: Hugo <hugo.talbot@sofa-framework.org>

* make consistent the info to warn change

* Apply suggestions from code review (clarifications in msg)

Co-authored-by: Hugo <hugo.talbot@sofa-framework.org>

* comments: correct english and replace container with MechanicalObject

Co-authored-by: Hugo <hugo.talbot@sofa-framework.org>

* comments correction

Co-authored-by: Hugo <hugo.talbot@sofa-framework.org>

* Update Component/Mass/src/sofa/component/mass/MeshMatrixMass.inl

Co-authored-by: Hugo <hugo.talbot@sofa-framework.org>

* Update Component/Mass/src/sofa/component/mass/DiagonalMass.inl

Co-authored-by: Hugo <hugo.talbot@sofa-framework.org>

* explicit links to topology and (geo) state

Co-authored-by: Alex Bilger <alxbilger@users.noreply.github.com>
Co-authored-by: Hugo <hugo.talbot@sofa-framework.org>

alxbilger committed:
[Sofa.LinearAlgebra] Test CRS product compared to Eigen (#2839)

* [Sofa.LinearAlgebra] Test CRS compared to Eigen

Also a demonstration on how to use Eigen::Map from a CRS matrix

* Avoid repetition

fredroy committed:
fix missing find_package for the generated cmake (#2841)

maddin200 committed:
[applications] fix shadow variable (#2844)

fredroy committed:
create fem.damping and move dampingff into it (+compat,etc)

alxbilger committed:
[examples] Optimal RequiredPlugin (#2836)

epernod committed:
[Scenes] Move PointSplatModel and SpatialGridPointModel scenes into Sph plugin as they are dedicated to it. And create a simpler version of PointSplatModel scene (#2847)

fredroy committed:
[SofaNG] Setup Constraint (#2790)

* create module constraint and submodule lagrangian

* create submodule projective

* add compat and redirection

* update component change

* update sofacuda

* split lagrangian between model,correction and solver

* update related files (compat, componentchange, cmake files...)

* use more precise packages for sofaconstraint

epernod committed:
[FEM.Elastic] Minor changes in TetraXX FEM at init and to be able to access Data  (#2845)

* [SofaMiscFem] Turn FastTetrahedralCorotational  Data as public

* [src] Update init to send warning on Data required

* [SofaSimpleFem] Add several getters

alxbilger committed:
[SofaNewMat] Fix compilation (#2829)

fredroy committed:
[SofaNG] Setup AnimationLoop (#2797)

* move file and clean

* update component change

* add compat layer

* fix rebase

fredroy committed:
Revert "create fem.damping and move dampingff into it (+compat,etc)"

This reverts commit 3b0efd14970ca4a5740c41fecfc88b8fb1c64e7c.

alxbilger committed:
[HyperElastic] Introduce example scene for StandardTetrahedralFEMForceField (#2857)

This is the same scene than TetrahedronHyperelasticityFEMForceField.scn but TetrahedronHyperelasticityFEMForceField has been replaced by StandardTetrahedralFEMForceField

dessteph committed:
[SofaSparseSolver] SparseLUSolver and SparseCholeskySolver support fill reducing permutation (#2788)

* [SofaSparseSolver]Added the choice to compute the decomposition at each step

* [SofaSparseSolver] Added the choice to apply fill-in reducing permutation

* Resolve conflicts

* [SparseSolver]Moved the if, avoid usless computation

* Fixed indentation

* Added timers for benchmark

* [SofaSparseSolver]SparseLUSolver and SparseCholeskySolver now support fill reducing permutation

* Better indentation

* Better timers

* Isolated the conversion from CSR to adj

* Use vectors intead of pointers

* Changed timers

* Isolated computation of adjancy matrix from csr format

* Deleted commented code

* Deleted useless code

* Update modules/SofaSparseSolver/src/SofaSparseSolver/SparseCholeskySolver.cpp

Co-authored-by: Alex Bilger <alxbilger@users.noreply.github.com>

* Update modules/SofaSparseSolver/src/SofaSparseSolver/SparseCholeskySolver.h

Co-authored-by: Alex Bilger <alxbilger@users.noreply.github.com>

* Changed NULL by nullptr, changed the include

* Update MatrixAssembly_direct_blocs.scn

* Added a boolean to compute the permutation only once (LU,Chol)

* Update CSR_to_adj.cpp

* Changed a path

* Apply a partial pivoy before solving (LU)

* Update modules/SofaSparseSolver/src/SofaSparseSolver/SparseLUSolver.h

Co-authored-by: Alex Bilger <alxbilger@users.noreply.github.com>

* Fix after merge from master

* WIP

* WIP

* Allow to choose between SuiteSparse and METIS

* Update SparseCholeskySolver.inl

* Avoid a copy loop

* WIP

* Cleaning, WIP

* Does not compute the parmutations at each step

* WIP

* Store the shape of the matrix only when needed

* Cleaning

* WIP

* Cleaning

* WIP

* Finalizing

* WIP

* WIP

* Fixed the SparseCholeskySolver without permutaions

* Fix

* Cleaning

* Cleaning

* Update SparseCommon.h

* Cleaning

Co-authored-by: Alex Bilger <alxbilger@users.noreply.github.com>
Co-authored-by: Alex Bilger <alexbilger0@gmail.com>

epernod committed:
[StateContainer] Remove deprecated code for topologicalChanges. Need to work on it.

alxbilger committed:
[SofaCUDA] Fix compilation for SOFA_GPU_CUDA_DOUBLE (#2863)

fredroy committed:
[Sofa.GL.Component] Fix tangents/bitangents type set for OpenGL (#2855)

epernod committed:
[SofaSimpleFEM_test] Add tests on Tetrahedron, CorotationalTetrahedral and FastTetrahedral FEM ForceField (#2842)

* [SofaSimpleFem] Add new tests for  TetrahedronFEMForceField / corotational and FastCorotational to check creation

* [SofaSimpleFEM] Add all tests for TetrahedronFEM, TetrahedralFEM and FastTetrahedral. Some are not implemented, others are failing.

* [src] up tests

* [tests] Setup basic tests on component behavior

* [SofaSimpleFem_test] Add test on TetrahedronFEM init

* [tests] Add test on init for TriangularFEM

* [tests] backup work on tetra tests

* [tests] Fix tetrahedralCorotationalFEM init

* [tests] Add tetrahedralCorotational and TetrahedronFEM values during simulation

* Fix compilation

* [SofaSimpleFem_test] Test vertex position after 100 step of Tetrahedron and tetrahedralCorotational

* [SofaSimpleFem_test] Add init and value tests for fastTetrahedralCorotationalFEM

* fix compilation

* Set class TetrahedronRestInformation definition public

* [tests] Update default values expected. Need to unify this value with other tetrahedronFEM

fredroy commented:
[ci-build][with-all-tests]

hugtalbot committed:
[SofaGuiQt] Remove sec unit from GUI (#2742)

* Remove sec unit from GUI

* Remove s

* Add step and use only frameCounter variable

* Update modules/SofaGuiQt/src/sofa/gui/qt/RealGUI.cpp

Co-authored-by: erik pernod <erik.pernod@gmail.com>

Co-authored-by: Frederick Roy <fredroy@users.noreply.github.com>
Co-authored-by: erik pernod <erik.pernod@gmail.com>

EulalieCoevoet committed:
[ArticulatedSystemPlugin] ArticulatedSystemMapping new features (#2803)

* [ArticulatedSystemPlugin] ArticulatedSystemMapping: new features
- allows to set the index for input2 instead of using the last one
- allows to set the container

* Update applications/plugins/ArticulatedSystemPlugin/src/ArticulatedSystemPlugin/ArticulatedSystemMapping.h

Co-authored-by: erik pernod <erik.pernod@gmail.com>

Co-authored-by: Frederick Roy <fredroy@users.noreply.github.com>
Co-authored-by: erik pernod <erik.pernod@gmail.com>

alxbilger committed:
[SofaCore] Name dynamically allocated vec ids (#2811)

* [SofaCore] Name dynamically allocated vec ids

* VecIds can be organized in groups

alxbilger committed:
[HyperElastic] Add draw method to StandardTetrahedralFEMForceField (#2838)

It is made common with TetrahedronHyperelasticityFEMForceField

epernod committed:
[Geomagic] Clean scene and code warnings + minor changes (#2846)

* [Geomagic]  Fix warnings due to pluginization of io::mesh

* [Geomagic] Update all scenes to remove warnings and optimize some physical parameters

* [Geomagic] Add 2 new scenes to test geomagic against a rigid and a deformable sphere

* [Geomagic] Clean some unused code

* Update GeomagicEmulator.cpp

fredroy committed:
move requiredplugin to simulationcore (with its test + compat) (#2849)

alxbilger committed:
[SofaBoundaryCondition] Remove optimization based on type of matrix (#2859)

alxbilger committed:
[StateContainer] Fix VecId names for != V_COORD

alxbilger committed:
[examples] Run PluginFinder after some module changes (#2860)

* Run PluginFinder after some module changes

* Revert cuda file

alxbilger committed:
[all] Move addInput/addOutput from init() to constructor (#2825)

* [all] Move addInput/addOutput from init() to constructor

* Revert move in SubsetTopology

alxbilger committed:
[SceneUtility] MakeDataAliasComponent: Remove alias in destructor (#2832)

* [SceneUtility] Remove alias in destructor

Due to the static nature of the ObjectFactory, the alias survived the destruction of a MakeDataAliasComponent. It is a problem if the program continues and still uses the ObjectFactory.

* Remove unused variable

* Remove unnecessary Data d_keepAliasAfterDestruction

* Private instead of protected

Co-authored-by: erik pernod <erik.pernod@gmail.com>

Co-authored-by: erik pernod <erik.pernod@gmail.com>

alxbilger committed:
[FEM.HyperElastic] Remove optimization based on type of matrix in StandardTetrahedralFEMForceField (#2858)

* [HyperElastic] Remove optimization based on type of matrix

* Fix return type

Co-authored-by: erik pernod <erik.pernod@gmail.com>

Co-authored-by: erik pernod <erik.pernod@gmail.com>

alxbilger committed:
[Metis] Disable compilation warnings (#2874)

ziqiu-zeng committed:
fix bug in linearsolvers and preconditioners, enable compilation with cuda plugins (#2883)

alxbilger commented on applications/plugins/ArticulatedSystemPlugin/examples/ArticulatedArm/header.py:
```diff
@@ -0,0 +1,22 @@
+
+    rootNode.addObject('VisualStyle', displayFlags='showVisualModels hideBehaviorModels hideForceFields hideWireframe')
+    rootNode.addObject('RequiredPlugin', name='SofaPlugins', pluginName=['ArticulatedSystemPlugin',
+     'SofaGeneralAnimationLoop', 'SofaPython3', 'SofaSparseSolver', 'SofaPreconditioner', 'SuctionCup',
```
`SuctionCup` cannot be loaded

EulalieCoevoet replied:
Oops, forgot to remove that.

guparan commented:
Scene test errors "extension (py) not handled" should be fixed now.

EulalieCoevoet committed:
[ArticulatedSystem] adds forgotten changes and cleans RequiredPlugin

EulalieCoevoet committed:
Merge branch 'master' into pr-addArticulatedSystemExample

EulalieCoevoet commented:
Okay I don't know why things appear like that. I think I just merged master into this branch. If someone knows please tell me. 

Anyway, I've cleaned the RequiredPlugin and added some forgotten changes.

alxbilger commented:
Still some problems with the scene on the CI, probably because tkinter is not installed.

```
Python exception:   
  ModuleNotFoundError: No module named 'tkinter'
```

tkinter must be removed from the scene OR this scene should be ignored by the CI.

EulalieCoevoet committed:
[ArticulatedSystem] cleaning

EulalieCoevoet commented:
The GUI (which requires tkinter) is optional but really nice I think to interact with the arm. I've just disabled it by default, so if you go for making the CI ignoring a file, it should only be the script `robotGUI.py`.

EulalieCoevoet committed:
[ArticulatedSystem] adds .scene-tests file (ignore robotGUI.py)

alxbilger commented:
@guparan @hugtalbot the scene has not been ignored despite the file .scene-tests

fredroy commented:
I think it is because it is missing the relative path iteself, i.e it should be "ArticulatedArm/robotGUI.py"
I will modify it directly if I can

fredroy committed:
add relative path for ignored test

damienmarchal commented:
Ready, so merge.
